azurerm/resource_arm_virtual_network_gateway_connection.go

```diff
@@ -0,0 +1,350 @@
+package azurerm
+
+import (
+	"fmt"
+	"github.com/Azure/azure-sdk-for-go/arm/network"
+	"github.com/hashicorp/errwrap"
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/hashicorp/terraform/helper/validation"
+	"log"
+	"net/http"
+	"time"
+)
+
+func resourceArmVirtualNetworkGatewayConnection() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceArmVirtualNetworkGatewayConnectionCreateUpdate,
+		Read:   resourceArmVirtualNetworkGatewayConnectionRead,
+		Update: resourceArmVirtualNetworkGatewayConnectionCreateUpdate,
+		Delete: resourceArmVirtualNetworkGatewayConnectionDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"resource_group_name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"type": {
+				Type:     schema.TypeString,
+				Required: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					string(network.ExpressRoute),
+					string(network.IPsec),
+					string(network.Vnet2Vnet),
+				}, false),
+				ForceNew: true,
+			},
+
+			"virtual_network_gateway_id": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"authorization_key": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+
+			"express_route_circuit_id": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+
+			"peer_virtual_network_gateway_id": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+
+			"local_network_gateway_id": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+
+			"enable_bgp": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  false,
+			},
+
+			"routing_weight": {
+				Type:     schema.TypeInt,
+				Optional: true,
+				Default:  10,
+			},
+
+			"shared_key": {
+				Type:      schema.TypeString,
+				Optional:  true,
+				Sensitive: true,
```

Comment:
I made it optional, since Site-to-Site and VNet-to-VNet connections require the key but not ExpressRoute connections. To clarify, I added to the field in the documentation "A key must be provided if a Site-to-Site or VNet-to-VNet connection is created whereas ExpressRoute connections do not need a shared key."

A more elegant way for validation would be if individual schema validator functions would have access to the entire resource to achieve "conditional mandatory" fields (see comments above).